### PR TITLE
New: Markdown Rendering for AI Crawlers (AEO) - Infra

### PR DESCRIPTION
**Pull request by matipojo**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
- [x] Feature

## Motivation

More and more people get their answers from AI — ChatGPT, Claude, Gemini, Perplexity - instead of clicking through search results. 

For website owners, this means their content needs to be understood by AI systems, not just rendered in a browser. Today, Elementor pages are wrapped in deeply nested HTML with layout divs, inline styles, and scripts that AI systems struggle to parse. Worse, content inside accordions, tabs, and toggles is hidden behind JavaScript interactions and may be completely invisible to AI crawlers.

Website owners who invest in great content deserve to have it surface in AI-generated answers. The Generative Engine Optimization (GEO) movement points the same way: sites that serve clean, structured Markdown get better AI visibility. This feature lets Elementor users opt in to that advantage - their pages automatically serve semantically accurate Markdown built from widget-level knowledge, not generic HTML stripping — so every heading, image, video, FAQ, and hidden panel is properly represented to AI systems.

See also:
- https://developers.cloudflare.com/fundamentals/reference/markdown-for-agents/
- https://vercel.com/blog/making-agent-friendly-pages-with-content-negotiation

## Example: Before & After

<details>
<summary><b>Without this feature</b> — what AI crawlers see today (raw HTML)</summary>

```html
<div data-elementor-type="wp-page" data-elementor-id="1759" class="elementor elementor-1759">
  <div class="elementor-element elementor-element-3c6e013 e-con-full e-flex e-con e-parent"
       data-id="3c6e013" data-element_type="container" data-e-type="container"
       data-settings="{&quot;position&quot;:&quot;fixed&quot;,&quot;background_background&quot;:&quot;classic&quot;}">
    <div class="elementor-element elementor-element-4c0b664 elementor-widget elementor-widget-heading"
         data-id="4c0b664" data-element_type="widget" data-widget_type="heading.default">
      <h4 class="elementor-heading-title elementor-size-default">🤖 NexaRobotics</h4>
    </div>
    <div class="elementor-element elementor-element-7e3498e elementor-widget elementor-widget-text-editor"
         data-id="7e3498e" data-element_type="widget" data-widget_type="text-editor.default">
      <p style="color: #ccc; font-size: 15px; font-weight: 500;">
        <a href="http://localhost/products" data-wplink-url-error="true">Products</a>
        <a href="http://localhost/Technology" data-wplink-url-error="true">Technology</a>
        <a href="http://localhost/About" data-wplink-url-error="true">About</a>
        <a href="http://localhost/Careers" data-wplink-url-error="true">Careers</a>
        <a href="http://localhost/Contact Us" data-wplink-url-error="true">
          <span style="color: #00d4ff; font-weight: bold;">Contact Us</span></a>
      </p>
    </div>
  </div>
  <!-- ... 800+ more lines of nested divs, data attributes, inline styles ... -->
</div>
```

</details>

<details>
<summary><b>With this feature</b> — clean Markdown served to AI crawlers</summary>

```markdown
---
title: "🤖 NexaRobotics"
description: "Next-generation autonomous robotics powered by AI. 500+ robots deployed worldwide with 99.8% uptime and enterprise-grade performance. "
url: "http://localhost:4444/elementor-page-1759"
date_modified: "2026-02-20T14:14:58+00:00"
---

#### 🤖 NexaRobotics

[Products](http://localhost/products)    [Technology](http://localhost/Technology)    [About](http://localhost/About)    [Careers](http://localhost/Careers)    [Contact Us](http://localhost/Contact Us)

---

###### ⚡ NEXT-GENERATION ROBOTICS

# The Future of Autonomous Robotics Starts Here

We build intelligent robotic systems that redefine what’s possible — from autonomous warehouses to precision surgical assistants. Powered by AI. Built for scale.

[Explore Our Robots](#)

[Watch Demo →](#)

---

## 500+

###### Robots Deployed

## 99.8%

...

```

</details>

## PR content

This is the **infrastructure PR** — it ships the full rendering pipeline, request handling, caching, and editor UI. Widget-specific `render_markdown()` overrides will follow in a separate PR on top of this one.

Even without widget overrides, every widget already produces simple text via the `widget-base.php` fallback (`wp_strip_all_tags` on rendered HTML).

Widget overrides are ready in #34842 (stacked on this PR).

- **Automatic Markdown responses** via `Accept: text/markdown` header or `?format=markdown` — zero user action beyond enabling the experiment
- **Experiment gating** (default inactive, alpha) under Elementor > Settings > Features
- **Cache TTL setting** under Settings > Performance
- **"View as Markdown"** menu item in the Editor top bar — preview what AI crawlers see
- **HTML→Markdown converter** — regex-based converter that handles headings, bold, italic, links, images, lists, blockquotes, code, and horizontal rules
- **Recursive element tree walker** — `render_markdown()` on `Element_Base` walks children; `Widget_Base` provides the strip-tags fallback

### What changed

| Area | Files | What |
|------|-------|------|
| **Core module** | `modules/markdown-render/` (3 files) | Experiment registration, request interception, caching, Markdown renderer, HTML→MD converter |
| **Base classes** | `element-base.php`, `widget-base.php` | `render_markdown()` — recursive traversal + `strip_tags` fallback |
| **Module registration** | `modules-manager.php` | Register `markdown-render` module |
| **Editor UI** | `packages/.../editor-app-bar/` (2 files) | "View as Markdown" menu item in the publish dropdown |
| **Document sync** | `packages/.../editor-documents/` (5 files) | `wpPreview` link exposed in document sync for the editor UI |
| **Tests** | `tests/.../markdown-render/` (2 files, 269 lines) | HTML→MD converter tests, module integration tests |

## Test instructions

1. Activate "Markdown Rendering" experiment in **Elementor > Settings > Features**
2. Publish a page with various widgets, open the editor, and use **"View as Markdown"** from the top-bar "Publish" menu
3. Verify the Markdown output contains the page content (headings, text, images, etc.)
4. Verify `curl -H "Accept: text/markdown" <page-url>` returns Markdown with YAML frontmatter
5. Confirm non-published / password-protected pages return normal 404
6. Confirm non-Elementor pages return normal HTML
7. Verify caching works — second request should be served from post meta cache

## Quality assurance

- [x] I have tested this code to the best of my abilities
- [x] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

## 1. Problem & Context

Adds infrastructure to serve Elementor page content as Markdown for AI crawlers (AEO) via Accept headers or ?format=markdown query param. Includes HTML→Markdown conversion pipeline and per-post caching with TTL configuration.

## 2. What Changed (Where)

- **Core**: Registered `markdown-render` module in Modules_Manager
- **Rendering**: Added `render_markdown()` methods to Element_Base and Widget_Base for recursive element traversal
- **Conversion**: New Html_To_Markdown class handles comprehensive HTML→Markdown transpilation (headings, links, lists, code blocks, etc.)
- **Orchestration**: Markdown_Renderer builds YAML frontmatter + element content; Module intercepts markdown requests with caching
- **Utilities**: Added `html_to_plain_text()` helper for tag stripping
- **UI/Frontend**: New "View as Markdown" button in editor, document preview link tracking (wpPreview)
- **Tests**: 26 unit tests covering conversion edge cases, YAML escaping, caching, request detection

## 3. How It Works

Request → Module detects `?format=markdown` or `Accept: text/markdown` header → checks cache (24h default TTL) → renders doc via Markdown_Renderer → builds frontmatter (title, date, featured_image, permalink) + recursively calls render_markdown() on all elements → Html_To_Markdown converts captured HTML to Markdown → returns as `text/markdown` response. Preview mode skips cache. Cache invalidates on post save, plugin activation/deactivation, theme switch.

## 4. Risks

**Regex DOS potential**: Multiple preg_replace passes on untrusted HTML could be slow on pathological inputs. Mitigate with content-length guards if needed. **Cache stampede**: High-traffic sites with expired cache could spike load during concurrent requests. Accept—caching is optional (configurable to disable). **Preview nonce validation**: Relies on existing WordPress preview nonce; ensure it's validated before frontmatter exposes preview URLs.

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using.
💡 **Tip:** You can customize your AI Description using **Guidelines** [Learn how](https://docs.gitstream.cm/automation-actions/#describe-changes)</sub>

### CODE REVIEW

File: modules/markdown-render/html-to-markdown.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
+<?php
+namespace Elementor\Modules\MarkdownRender;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+class Html_To_Markdown {
+
+	public static function convert( string $html ): string {
```
Comment (Ntnelbaba) [inline modules/markdown-render/html-to-markdown.php:10]: Something weird here
There is a convert function (and a full list of tools), but it is actually not being used
Why is this class needed?

File: modules/markdown-render/markdown-renderer.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+			$lines[] = 'description: "' . $this->escape_yaml_string( $description ) . '"';
+		}
+
+		$thumbnail = get_the_post_thumbnail_url( $post_id, 'full' );
+
+		if ( $thumbnail ) {
+			$lines[] = 'featured_image: "' . esc_url( $thumbnail ) . '"';
+		}
+
+		$lines[] = 'url: "' . get_permalink( $post_id ) . '"';
```
Comment (Ntnelbaba) [inline modules/markdown-render/markdown-renderer.php]: get_permalink might resolve to false, we need to validate before adding this line

File: modules/markdown-render/html-to-markdown.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+			'#<a[^>]+href=["\']([^"\']*)["\'][^>]*>(.*?)</a>#is',
+			function ( $matches ) {
+				$url = $matches[1];
+				$text = $matches[2];
+
+				if ( empty( $url ) || '#' === $url ) {
+					return $text;
+				}
+
+				return '[' . $text . '](' . $url . ')';
```
Comment (Ntnelbaba) [inline modules/markdown-render/html-to-markdown.php:88]: The link text $text may contain `]` characters, breaking the generated `[text](url)` syntax. Example: `<a href="...">Click [here]</a>` → `[Click [here]](url)`

File: modules/markdown-render/html-to-markdown.php (hunk 1 above)
Comment (louiswol94) [reply modules/markdown-render/html-to-markdown.php:10]: It's used in the follow up PR in `includes/widgets/html.php`

File: modules/markdown-render/markdown-renderer.php (hunk 2 above)
Comment (louiswol94) [reply modules/markdown-render/markdown-renderer.php]: Same applies to `get_the_modified_date()`

File: modules/markdown-render/markdown-renderer.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		$markdown = $element->render_markdown();
+
+		return apply_filters( 'elementor/markdown/element_output', $markdown, $element, $element_data );
+	}
+
+	private function escape_yaml_string( string $value ): string {
+		$value = str_replace( "\xE2\x80\x8B", '', $value );
+
+		return str_replace( [ '"', "\n", "\r" ], [ '\\"', ' ', '' ], $value );
+	}
```
Comment (louiswol94) [inline modules/markdown-render/markdown-renderer.php:115]:
```suggestion
	private function escape_yaml_string( string $value ): string {
        $value = html_entity_decode( $value, ENT_QUOTES | ENT_HTML5, 'UTF-8' );
        $value = str_replace( "\xE2\x80\x8B", '', $value );
        $value = preg_replace( '/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/', '', $value );
    
        return strtr( $value, [
            '\\' => '\\\\',
            '"'  => '\\"',
            "\n" => '\\n',
            "\r" => '\\r',
            "\t" => '\\t',
        ] );
  }
```

**Review by louiswol94 — DISMISSED** [review]
Happy to ship this in alpha 🚀

File: modules/markdown-render/markdown-renderer.php (hunk 4 above)
Comment (matipojo) [reply modules/markdown-render/markdown-renderer.php:115]: Fixed

File: modules/markdown-render/html-to-markdown.php (hunk 3 above)
Comment (matipojo) [reply modules/markdown-render/html-to-markdown.php:88]: Fixed + test

File: modules/markdown-render/markdown-renderer.php (hunk 2 above)
Comment (matipojo) [reply modules/markdown-render/markdown-renderer.php]: Fixed + test

**Review by Ntnelbaba — APPROVED** [review]

**Review by louiswol94 — APPROVED** [review]
